File: client/android/GoogleMapsA2UI/src/main/java/com/google/android/libraries/mapsplatform/a2ui/A2AResponseParser.kt
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.android.libraries.mapsplatform.a2ui

import org.json.JSONArray
import org.json.JSONObject

sealed class A2AParserException(message: String) : Exception(message) {
  class InvalidPayloadStructure(
    message: String = "The raw JSON response lacks a recognized message parts structure."
  ) : A2AParserException(message)
}

data class ParsedA2AEventMetadata(val mimeType: String?)

sealed interface ParsedA2AEvent {
  data class Text(val text: String) : ParsedA2AEvent

  data class Data(val data: String, val metadata: ParsedA2AEventMetadata? = null) : ParsedA2AEvent
}

object A2AResponseParser {

  private const val KEY_TEXT = "text"
  private const val KEY_KIND = "kind"
  private const val KEY_DATA = "data"
  private const val KEY_HISTORY = "history"
  private const val KEY_ROLE = "role"
  private const val VAL_USER = "user"
  private const val VAL_AGENT = "agent"
  private const val KEY_PARTS = "parts"
  private const val KEY_CONTENT = "content"
  private const val KEY_STATUS = "status"
  private const val KEY_MESSAGE = "message"

  private const val KEY_CREATE_SURFACE = "createSurface"
  private const val KEY_UPDATE_COMPONENTS = "updateComponents"
  private const val KEY_UPDATE_DATA_MODEL = "updateDataModel"
  private const val KEY_DELETE_SURFACE = "deleteSurface"
  private const val KEY_SURFACE_ID = "surfaceId"

  private val A2UI_PATTERN = "<a2ui-json>(.*?)</a2ui-json>".toRegex(RegexOption.DOT_MATCHES_ALL)

  fun parse(rawJson: JSONObject): List<ParsedA2AEvent> {
    val partsArray =
      extractPartsArray(rawJson) ?: throw A2AParserException.InvalidPayloadStructure()

    val partsList = mutableListOf<ParsedA2AEvent>()
    var currentTextBuilder = java.lang.StringBuilder()
    var currentUiElements = JSONArray()

    for (i in 0 until partsArray.length()) {
      val part = partsArray.getJSONObject(i)
      val textPart =
        if (part.has(KEY_TEXT)) part.optString(KEY_TEXT)
        else if (part.optString(KEY_KIND) == KEY_TEXT) part.optString(KEY_TEXT) else null

      if (textPart != null) {
        if (currentUiElements.length() > 0) {
          partsList.add(ParsedA2AEvent.Data(currentUiElements.toString()))
          currentUiElements = JSONArray()
        }

        if (textPart.contains("<a2ui-json>")) {
          extractJsonBlocks(textPart, currentTextBuilder, partsList)
        } else if (textPart.isNotEmpty()) {
          if (currentTextBuilder.isNotEmpty()) currentTextBuilder.append("\n")
          currentTextBuilder.append(textPart)
        }
      }

      val dataPayload =
        if (part.has(KEY_DATA)) part.optJSONObject(KEY_DATA)
        else if (part.optString(KEY_KIND) == KEY_DATA) part.optJSONObject(KEY_DATA) else null
      if (dataPayload != null && isUiElement(dataPayload)) {
        if (currentTextBuilder.isNotEmpty()) {
          partsList.add(ParsedA2AEvent.Text(currentTextBuilder.toString()))
          currentTextBuilder = java.lang.StringBuilder()
        }
        currentUiElements.put(dataPayload)
      }
    }

    if (currentTextBuilder.isNotEmpty()) {
      partsList.add(ParsedA2AEvent.Text(currentTextBuilder.toString()))
    }
    if (currentUiElements.length() > 0) {
      partsList.add(ParsedA2AEvent.Data(currentUiElements.toString()))
    }

    val deduplicatedParts = mutableListOf<ParsedA2AEvent>()
    val seenSurfaces = mutableSetOf<String>()
    var lastSeenText: String? = null

    for (part in partsList) {
      val finalText: String? =
        if (part is ParsedA2AEvent.Text) {
          part.text.trim().takeIf { it.isNotEmpty() }
        } else null

      // Deduplicate consecutive identical text blocks
      if (finalText != null && finalText != lastSeenText) {
        deduplicatedParts.add(ParsedA2AEvent.Text(finalText))
        lastSeenText = finalText
      }

      if (part is ParsedA2AEvent.Data && part.data != "[]") {
        try {
          val array = JSONArray(part.data)
          val newArray = JSONArray()
          for (i in 0 until array.length()) {
            val obj = array.getJSONObject(i)
            val sid = obj.optJSONObject(KEY_CREATE_SURFACE)?.optString(KEY_SURFACE_ID)
            if (sid != null) {
              if (seenSurfaces.contains(sid)) {
                continue
              }
              seenSurfaces.add(sid)
            }
            newArray.put(obj)
          }
          if (newArray.length() > 0) {
            deduplicatedParts.add(ParsedA2AEvent.Data(newArray.toString(), part.metadata))
          }
        } catch (e: Exception) {
          if (part.data.isNotEmpty()) {
            deduplicatedParts.add(part)
          }
        }
      }
    }

    return deduplicatedParts
  }

  private fun isUiElement(obj: JSONObject): Boolean {
    return obj.has(KEY_CREATE_SURFACE) ||
      obj.has(KEY_UPDATE_COMPONENTS) ||
      obj.has(KEY_UPDATE_DATA_MODEL) ||
      obj.has(KEY_DELETE_SURFACE)
  }

  private fun extractPartsArray(rawJson: JSONObject): JSONArray? {
    val finalParts = JSONArray()

    if (rawJson.has(KEY_HISTORY)) {
      val history = rawJson.optJSONArray(KEY_HISTORY)
      if (history != null) {
        var lastUserIndex = -1
        for (i in 0 until history.length()) {
          val msg = history.optJSONObject(i)
          if (msg?.optString(KEY_ROLE) == VAL_USER) {
            lastUserIndex = i
          }
        }

        for (i in (lastUserIndex + 1) until history.length()) {
          val msg = history.optJSONObject(i)
          if (msg?.optString(KEY_ROLE) == VAL_AGENT) {
            val parts = msg.optJSONArray(KEY_PARTS)
            if (parts != null) {
              for (j in 0 until parts.length()) {
                finalParts.put(parts.getJSONObject(j))
              }
            }
          }
        }
      }
    }

    val additionalParts =
      when {
        rawJson.has(KEY_PARTS) -> rawJson.optJSONArray(KEY_PARTS)
        rawJson.has(KEY_CONTENT) -> rawJson.optJSONObject(KEY_CONTENT)?.optJSONArray(KEY_PARTS)
        rawJson.has(KEY_STATUS) ->
          rawJson.optJSONObject(KEY_STATUS)?.optJSONObject(KEY_MESSAGE)?.optJSONArray(KEY_PARTS)
        else -> null
      }

    if (additionalParts != null) {
      for (i in 0 until additionalParts.length()) {
        finalParts.put(additionalParts.getJSONObject(i))
      }
    }

    val hasRecognizedStructure =
      rawJson.optJSONArray(KEY_HISTORY) != null || additionalParts != null
    return if (hasRecognizedStructure) finalParts else null
  }

  private fun extractJsonBlocks(
    textPart: String,
    textBuilder: java.lang.StringBuilder,
    partsList: MutableList<ParsedA2AEvent>,
  ) {
    val allMatches = A2UI_PATTERN.findAll(textPart).toList()

    if (allMatches.isEmpty()) {
      if (textBuilder.isNotEmpty()) textBuilder.append("\n")
      textBuilder.append(textPart)
      return
    }

    var lastEnd = 0
    for (match in allMatches) {
      val beforeText = textPart.substring(lastEnd, match.range.first).trim()
      if (beforeText.isNotEmpty()) {
        if (textBuilder.isNotEmpty()) textBuilder.append("\n")
        textBuilder.append(beforeText)
      }

      val jsonString = match.groupValues[1].trim()
      try {
        val firstChar = jsonString.firstOrNull()
        if (firstChar == '[') {
          val array = JSONArray(jsonString)
          val localUiElements = JSONArray()
          for (i in 0 until array.length()) {
            localUiElements.put(array.getJSONObject(i))
          }
          if (textBuilder.isNotEmpty()) {
            partsList.add(ParsedA2AEvent.Text(textBuilder.toString()))
            textBuilder.clear()
          }
          partsList.add(ParsedA2AEvent.Data(localUiElements.toString()))
        } else if (firstChar == '{') {
          val jsonObj = JSONObject(jsonString)
          if (textBuilder.isNotEmpty()) {
            partsList.add(ParsedA2AEvent.Text(textBuilder.toString()))
            textBuilder.clear()
          }
          val localUiElements = JSONArray()
          localUiElements.put(jsonObj)
          partsList.add(ParsedA2AEvent.Data(localUiElements.toString()))
        } else {
          if (textBuilder.isNotEmpty()) textBuilder.append("\n")
          textBuilder.append(match.value)
        }
      } catch (e: Exception) {
        if (textBuilder.isNotEmpty()) textBuilder.append("\n")
        textBuilder.append(match.value)
      }
      lastEnd = match.range.last + 1
    }

    val remainingText = textPart.substring(lastEnd).trim()
    if (remainingText.isNotEmpty()) {
      if (textBuilder.isNotEmpty()) textBuilder.append("\n")
      textBuilder.append(remainingText)
    }
  }
}

File: client/ios/GoogleMapsA2UI/Sources/GoogleMapsA2UI/A2AResponseParser.swift
//
// Copyright 2026 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import Foundation

/// Errors that can occur during A2UI response parsing.
public enum A2AParserError: Error {
  /// The raw JSON response lacks a recognized message parts structure.
  case invalidPayloadStructure
  /// The provided dictionary cannot be serialized as valid JSON.
  case invalidJSONFormat
}

/// Utility parser to process server JSON responses.
/// Declared as a case-less enum to prevent instantiation.
public enum A2AResponseParser {
  private static let a2uiJsonMimeType = "application/json+a2ui"
  private static let a2uiJsonTagOpen = "<a2ui-json>"
  private static let a2uiJsonTagClose = "</a2ui-json>"

  /// Parses a raw server response dictionary into a flat list of `ParsedA2AEvent`s.
  ///
  /// The parser checks multiple possible JSON paths (`parts`, `content.parts`, `status.message.parts`) 
  /// to support varying response structures from different server backends (e.g. standalone JSON-RPC vs. ADK Web Server).
  ///
  /// Note: Any extracted A2UI JSON payloads (mime type `application/json+a2ui`) will always be batched 
  /// and returned as an array (`[Any]`) inside `ParsedA2AEvent.data`, providing a consistent format.
  ///
  /// - Parameter rawJSON: The raw JSON dictionary received from the server.
  /// - Returns: An array of `ParsedA2AEvent` objects extracted from the payload.
  /// - Throws: `A2AParserError.invalidJSONFormat` if the input is not valid JSON, or `A2AParserError.invalidPayloadStructure` if the parts array cannot be found.
  public static func parse(_ rawJSON: [String: Any]) throws -> [ParsedA2AEvent] {
    guard JSONSerialization.isValidJSONObject(rawJSON) else {
      throw A2AParserError.invalidJSONFormat
    }

    let partsArray: [[String: Any]]?
    if let parts = rawJSON["parts"] as? [[String: Any]] {
      partsArray = parts
    } else if let content = rawJSON["content"] as? [String: Any],
              let parts = content["parts"] as? [[String: Any]] {
      partsArray = parts
    } else if let status = rawJSON["status"] as? [String: Any],
              let message = status["message"] as? [String: Any],
              let parts = message["parts"] as? [[String: Any]] {
      partsArray = parts
    } else {
      partsArray = nil
    }

    guard let parts = partsArray else {
      throw A2AParserError.invalidPayloadStructure
    }

    var sdkParts: [ParsedA2AEvent] = []
    var a2uiPayloads: [Any] = []

    // flushA2UI() batches consecutive A2UI JSON payloads together into a single ParsedA2AEvent. 
    func flushA2UI() {
      if !a2uiPayloads.isEmpty {
        let event = ParsedA2AEvent.data(
          a2uiPayloads,
          metadata: ParsedA2AEventMetadata(mimeType: a2uiJsonMimeType)
        )
        sdkParts.append(event)
        a2uiPayloads.removeAll()
      }
    }

    for part in parts {
      if let textPart = part["text"] as? String {
        flushA2UI()
        let subEvents = splitTextParts(textPart)
        sdkParts.append(contentsOf: subEvents)
      } else if let dataPayload = part["data"] as? [String: Any] {
        let metadata = part["metadata"] as? [String: Any]
        let mimeType = part["mimeType"] as? String ?? metadata?["mimeType"] as? String
        let resolvedMimeType = mimeType
          ?? (isA2UIPayload(dataPayload) ? a2uiJsonMimeType : nil)

        if resolvedMimeType == a2uiJsonMimeType {
          a2uiPayloads.append(dataPayload)
        } else {
          flushA2UI()
          let event = ParsedA2AEvent.data(
            dataPayload,
            metadata: ParsedA2AEventMetadata(mimeType: resolvedMimeType)
          )
          sdkParts.append(event)
        }
      }
    }

    flushA2UI()

    return sdkParts
  }

  private static let a2uiKeys: Set<String> = [
    "createSurface", "updateComponents", "updateDataModel",
    "beginRendering", "surfaceUpdate", "dataModelUpdate"
  ]

  /// Checks if a given dictionary represents an A2UI payload.
  ///
  /// - Parameter dict: The dictionary to check.
  /// - Returns: `true` if the dictionary contains recognizable A2UI keys; otherwise, `false`.
  private static func isA2UIPayload(_ dict: [String: Any]) -> Bool {
    return dict.keys.contains { a2uiKeys.contains($0) }
  }

  /// Parses a JSON string into a dictionary or array.
  ///
  /// - Parameter jsonStr: The JSON string to parse.
  /// - Returns: The parsed JSON object, or `nil` if parsing fails.
  private static func parseJSON(_ jsonStr: String) -> Any? {
    guard let data = jsonStr.data(using: .utf8) else { return nil }
    return try? JSONSerialization.jsonObject(with: data, options: [])
  }

  /// Splits a text part containing embedded `<a2ui-json>` tags into separate events.
  ///
  /// - Parameter textPart: The string containing text and potentially embedded JSON.
  /// - Returns: An array of `ParsedA2AEvent`s representing the separated text and data parts.
  private static func splitTextParts(_ textPart: String) -> [ParsedA2AEvent] {
    var parts: [ParsedA2AEvent] = []
    if textPart.contains(a2uiJsonTagOpen) {
      var remainingText = textPart
      while let startRange = remainingText.range(of: a2uiJsonTagOpen) {
        let intro = String(remainingText[..<startRange.lowerBound])
          .trimmingCharacters(in: .whitespacesAndNewlines)
        if !intro.isEmpty {
          parts.append(.text(intro))
        }

        let rest = remainingText[startRange.upperBound...]
        let jsonStr: String
        if let endRange = rest.range(of: a2uiJsonTagClose) {
          jsonStr = String(rest[..<endRange.lowerBound])
          remainingText = String(rest[endRange.upperBound...])
        } else {
          jsonStr = String(rest)
          remainingText = ""
        }

        let trimmedJSON = jsonStr.trimmingCharacters(in: .whitespacesAndNewlines)
        let dataValue = parseJSON(trimmedJSON) ?? trimmedJSON
        let event = ParsedA2AEvent.data(
          [dataValue],
          metadata: ParsedA2AEventMetadata(mimeType: a2uiJsonMimeType)
        )
        parts.append(event)
      }
      let remaining = remainingText.trimmingCharacters(in: .whitespacesAndNewlines)
      if !remaining.isEmpty {
        parts.append(.text(remaining))
      }
    } else {
      if !textPart.isEmpty {
        parts.append(.text(textPart))
      }
    }
    return parts
  }
}
